### DISCUSSION TITLE
Svelte is for newbies but SvelteKit is not? (why I ditched SvelteKit)

### DISCUSSION THREAD

**Discussion by jtlapp** [sveltejs/kit#4586] [General]
Hello Svelties (what do we call ourselves?),

I found vanilla Svelte to be enjoyable and easy to learn, having already implemented [an Electron app with Svelte](https://github.com/ut-entomology/spectool), but I'm ready to ditch SvelteKit to go with an SPA website.

SvelteKit works fine out of the box, but I struggled with the following:

- Adding non-browser unit testing to test my backend code. I put at least two days into trying to figure this out and ultimately landed on [vitest-svelte-kit](https://github.com/nickbreaton/vitest-svelte-kit), which works great, but VSCode support is still less than ideal. See issue #4423.
- Adding CLI tools that call the backend code. The SvelteKit requirement that `"type": "module"` be in `package.json` seemed to be the underlying problem. I probably put half a day into trying to figure this out. @dominikg suggested a few ways to deal with this in discussion #4534, but read on...
- Using SvelteKit in a monorepo. I actually had no trouble setting up a basic SvelteKit installation in a monorepo, but I could not get a port of my site to compile in the monorepo. I faced problem after problem and am ready to quit, with [this slew of errors](https://github.com/sveltejs/kit/discussions/4534#discussioncomment-2531625) remaining (see also #4551). This is one full day's worth of effort so far with no end in sight.

I suspect that none of these reflect problems with SvelteKit itself, but they do suggest that the "Svelte is for newbies" (like Virginia is for lovers?) marketing does not translate to SvelteKit. If you want to make SvelteKit accessible to newbies, it may be necessary to either compile a list of especially helpful articles or else add best practices to the SvelteKit docs.

Mind you, I've written several express/handlebars websites over the years (even in TypeScript), so I'm not a complete web newbie. And I've been programming professionally for 35 years (mostly Java/C/C++), so I'm by no means a newbie developer. I do find the base knowledge needed for serious web development a bit mindboggling, though.

**Comment by jtlapp** [comment]
It took me 3 hours to convert my SvelteKit app into a Svelte SPA, including the following:

- Frontend and backend are building in separate directories with separate tsconfigs.
- I'm now running my unit tests with jest, giving me state-of-the-art VSCode support.
- My CLI tools now build and run.

That is, it took me 3 hours to do with vanilla Svelte what I failed to do in 28 hours with SvelteKit.

I may convert back to SvelteKit in a few months if you can provide better guidance.

**Comment by jtlapp** [comment]
Would providing an official monorepo template solve these problems? The template would need to make it easy to the following:

- Separate backend from frontend code.
- Make it easy to use jest for testing backend code.
- Demo building backend CLI tools.

**Comment by benmccann** [comment]
- you said [Vitest works perfectly](https://github.com/sveltejs/kit/issues/4423#issuecomment-1079000074) except for the VS Code extensions. However, the only issues I'm aware of that you've run into [have been fixed](https://github.com/zxch3n/vitest-explorer/issues/11). It's unclear what's left to do here, but if there are still problems that you're facing please file an issue. Otherwise we can't address it if we don't know about it
- the sourcemap issue seems to be fixed and just needing a new release of [`svelte-preprocess`](https://github.com/sveltejs/kit/discussions/4551#discussioncomment-2531965)
- a CLI is a very uncommon requirement, so I'm not sure it makes sense for us to document that specifically. There are a number of examples of SvelteKit running in a monorepo already such as [the sites repo](https://github.com/sveltejs/sites)

**Reply by jtlapp** [reply]
Thanks for looking into these a bit. My responses:

- The vitest VSCode extension has UI issues. It doesn't show the line where an error occurred, doesn't show where `expect` tests failed, and doesn't show the `expect` mismatch. The author is apparently outputting this information to the screen, but it's usually offscreen and not scrollable into view. I didn't create an issue for these because the author and I corresponded separately on them. I ultimately stopped using the extension and used `.only` to debug tests.
- Thanks for fixing the sourcemap issue. I wasn't sure that it wasn't still another problem with my configuration.
- I was dealing with [a number of odd errors](https://github.com/sveltejs/kit/discussions/4534#discussioncomment-2531625) associated with building/running CLI that shared code with my svelte-kit solution. I couldn't tell if this was due to incompatible tsconfig files or if I was somehow using pnpm wrong.

Unfortunately, I didn't keep my very broken pnpm repo, so I don't have a way to test your sourcemap fix. I also converted my svelte-kit solution to pure Svelte SPA. I plan to revisit SSR again in a few months, and I'm hoping I'll find something that inspires me to try again with svelte-kit.

**Reply by jtlapp** [reply]
I thought of a way that I might be able to return to a svelte-kit monorepo without having to fight incompatibility issues between front-end-and backend compilation: don't use svelte-kit endpoints at all, using svelte-kit only for the front end.
